### PR TITLE
Refactor of metadata plugin machinery

### LINKED ISSUE #67: Rethink plugin system and support
Currently we have a small set of hardcoded metadata plugins, that we recently merged with the main repo.

This system is not extensible and hard to configure, even in our environment. Think how can we improve this and open it a bit.

### PR DESCRIPTION
**Pull request by psrok1**
**Your checklist for this pull request**
- [x] I've read the [contributing guideline](https://github.com/CERT-Polska/mquery/blob/master/CONTRIBUTING.md).
- [ ] I've tested my changes by building and running mquery, and testing changed functionality (if applicable)
- [ ] I've added automated tests for my change (if applicable, optional)
- [ ] I've updated documentation to reflect my change (if applicable)

**What is the current behaviour?**
- Core plugin machinery (metadata) and plugins are contained in the same directory which could be difficult to maintain
- Cache need to be handled manually by plugin code

**What is the new behaviour?**
- Metadata base class renamed to MetadataPlugin abstract base class and moved from `metadata/__init__.py` to `src/metadata.py`. Now `src/metadata` is meant to be controlled by user.
- Added new overridable method: `identify` which allows to extract cache identifier from matched path and check whether match is applicable for plugin
- Refactored existing plugins
- Removed unused parts of `config.*.py`

**Test plan**

🤔 

Part of #67

### CODE REVIEW

File: src/metadata.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+Metadata = Dict[str, Any]
+
+
+class MetadataPlugin(ABC):
+    __depends_on__: List[str] = []
+    __cacheable__: bool = False
+    __cache_expire_time__: int = DEFAULT_CACHE_EXPIRE_TIME
+
+    def __init__(self):
```
Comment (nazywam) [inline src/metadata.py:15]:
```suggestion
    def __init__(self) -> None:
```

File: src/metadata.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        self.redis = None
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    def set_redis(self, redis):
```
Comment (nazywam) [inline src/metadata.py:22]:
```suggestion
    def set_redis(self, redis) -> None:
```

File: src/metadata.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    def _cache_fetch(self, cache_tag: str) -> Metadata:
+        rs_key = self.__rs_key(cache_tag)
+        obj = self.redis.get(rs_key)
+
+        if obj:
+            self.redis.expire(rs_key, self.__cache_expire_time__)
+            return json.loads(obj)
+        return {}
+
+    def _cache_store(self, cache_tag: str, obj: Metadata):
```
Comment (nazywam) [inline src/metadata.py:37]:
```suggestion
    def _cache_store(self, cache_tag: str, obj: Metadata) -> None:
```

File: src/metadata.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        rs_key = self.__rs_key(cache_tag)
+        self.redis.setex(rs_key, self.__cache_expire_time__, json.dumps(obj))
+
+    def identify(self, matched_fname: str) -> Optional[str]:
```
Comment (nazywam) [inline src/metadata.py:41]: Pretty interesting function 🤔

File: src/plugins/cuckoo_analysis.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import os
+import re
+
+from typing import Optional
+
+from metadata import Metadata, MetadataPlugin
+
+
+class CuckooAnalysisMetadata(MetadataPlugin):
+    def __init__(self, path):
```
Comment (nazywam) [inline src/plugins/cuckoo_analysis.py:10]:
```suggestion
    def __init__(self, path) -> None:
```

File: src/metadata.py (hunk 4 above)
Comment (nazywam) [reply src/metadata.py:41]: I guess it makes sense in an abstract class though

**Comment by psrok1** [comment]
For another PR:
- Deliver configuration to plugins in a more convenient manner
- Additional plugin requirements handling

Also not sure how to easily provide automatic tests for that.

File: src/metadata.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        if not self.redis:
+            return
+        rs_key = self.__rs_key(cache_tag)
+        self.redis.setex(rs_key, self.__cache_expire_time__, json.dumps(obj))
+
+    def identify(self, matched_fname: str) -> Optional[str]:
+        """
+       Returns file unique identifier based on matched path.
+
+       Intended to be overridden by plugin.
```
Comment (msm-code) [inline src/metadata.py:53]: OCD:
```suggestion
        Returns file unique identifier based on matched path.

        Intended to be overridden by plugin.
```

File: src/metadata.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        self.redis = redis
+
+    def __rs_key(self, cache_tag: str) -> str:
+        return f"cached:{self.name}:{cache_tag}"
+
+    def _cache_fetch(self, cache_tag: str) -> Metadata:
+        if not self.redis:
+            return {}
+        rs_key = self.__rs_key(cache_tag)
+        obj = self.redis.get(rs_key)
```
Comment (msm-code) [inline src/metadata.py:36]: Not for this PR, but redis queries can be costly, maybe we could use cachetools to speed it up later :thinking:

File: src/metadata.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+       Intended to be overridden by plugin.
+
+       :param identifier: File identifier returned by overridable
+                          :py:meth:`MetadataPlugin.identify` method
+       :param matched_fname: Matched file path
+       :param current_meta: Metadata extracted so far by dependencies
+       :return: Metadata object. If you can't extract metadata for current file,
+                return empty dict.
+       """
```
Comment (msm-code) [inline src/metadata.py:88]:
```suggestion
        Extracts metadata for matched path

        Intended to be overridden by plugin.

        :param identifier: File identifier returned by overridable
                           :py:meth:`MetadataPlugin.identify` method
        :param matched_fname: Matched file path
        :param current_meta: Metadata extracted so far by dependencies
        :return: Metadata object. If you can't extract metadata for current file,
                 return empty dict.
       """
```

File: src/metadata.py (hunk 7 above)
Comment (psrok1) [reply src/metadata.py:36]: Yup, maybe it's better to keep it locally.

**Pull request merged**